csi/controller.go

```diff
@@ -115,19 +118,122 @@ func (s *OsdCsiServer) ControllerGetCapabilities(
 // ControllerPublishVolume is a CSI API implements the attachment of a volume
 // on to a node.
 func (s *OsdCsiServer) ControllerPublishVolume(
-	context.Context,
-	*csi.ControllerPublishVolumeRequest,
+	ctx context.Context,
+	req *csi.ControllerPublishVolumeRequest,
 ) (*csi.ControllerPublishVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "This request is not supported")
+
+	logrus.Debugf("Calling Controller Publish Volume in the csi controller")
+	// check if volumeID is present and volume is available on SP
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "VolumeID is mandatory in ControllerPublishVolume request")
+	}
+	clogger.WithContext(ctx).Infof("ControllerPublishVolume request recived. VolumeID: %s", req.GetVolumeId())
+
+	// if nodeID is set, and node is actually present on SP
+	if req.NodeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeID is mandatory in CPVolume request")
+	}
+	// if VolumeCapability is nil, fail.
+	if req.VolumeCapability == nil {
+		return nil, status.Error(codes.InvalidArgument, "VolumeCaps must be present in the NodeSVolReq")
+	}
+	vol, err := s.driverGetVolume(ctx, req.GetVolumeId())
+	if err != nil {
+		if s, ok := status.FromError(err); ok && s.Code() == codes.NotFound {
+			return nil, status.Error(codes.NotFound, "Volume not found")
+		}
+		return nil, err
+	}
+
+	// have the volumeID, submit a new grpc request to px driver to prepare
+	// volume to be published only if the volume propery is winshare.
+	conn, err := s.getRemoteConn(ctx)
```

Comment:
This will get a remote conn on a random node in the cluster. What is the benefit of using ControllerPublishVolume if both NodePublish/ControllerPublish will pick a random node in the cluster?

The purpose of ControllerPublish is to have 2-3 nodes in a cluster that run the controller drivers. The node driver runs on every node in the cluster, as we do with Portworx currently.

Reply:
@ggriffiths I am not entirely convinced that the changes are understood here.

ControllerPublish will get called for every volume attachment in the k8s cluster inclusive of windows nodes where px does not run. The subsequent Node{Stage|Publish}... et al will get called on the scheduled windows node.
NodePublish is current implementation for volume attachment on linux nodes. 

> The purpose of ControllerPublish is to have 2-3 nodes in a cluster that run the controller drivers. 

not necessarily true given portworx csi driver is in itself an exception. This change is only an extension to supported capability to add ControllerPublish support.